### ISSUE TITLE
Attributes from devicesset in schematic is not picked up in the BOM

### ISSUE THREAD

**Issue by JacobChrist** [I2SE/eagle-bom.py#6]
For example the following attribute "PLP#" will be picked up from this part:
```xml
<part name="R13" library="PLP" deviceset="PLP-27" device="" value="10K">
<attribute name="PLP#" value="PLP-27"/>
</part>
```

However this part:

```xml
<part name="C2" library="PLP" deviceset="PLP-101" device="" value="10uF"/>
```

Will not pick up the attributes "PLP#" and "VALUE" in its deviceset:

```xml
<deviceset name="PLP-101" prefix="C" uservalue="yes">
<description>Capacitor 0805</description>
<gates>
<gate name="G$1" symbol="DISCRETE_CAP" x="0" y="0"/>
</gates>
<devices>
<device name="" package="DISCRETE_C0805">
<connects>
<connect gate="G$1" pin="+" pad="1"/>
<connect gate="G$1" pin="-" pad="2"/>
</connects>
<technologies>
<technology name="">
<attribute name="PLP#" value="PLP-101"/>
<attribute name="VALUE" value="10uF"/>
</technology>
</technologies>
</device>
</devices>
</deviceset>
```

I'm not sure what causes an attribute to be copied from the deviceset to the part, smashing a part will cause attributes to be copied from the deviceset to the instance (so positions on the schematic can be stored).  If you right click and edit an attribute for a part this will cause them to be copied (but only if modified).

It seems as the deviceset attributes should be added to the elements array first then overwritten by part attributes.

Jacob

**Comment by JacobChrist** [comment]
Then something like this:

Build an array of each unique deviceset.
```python
deviceset_find_string = "schematic/devicesets/deviceset"
for elem in drawing.iterfind(deviceset_find_string):
```
This in this loop:
```python
    for elem in drawing.iterfind(part_find_string):
```
when a part is found pull all attributes from the deviceset and add them to the elements.

**Comment by ch-aurich** [comment]
I am not sure if I understood your problem correct. Could you please provide a eagle file where the problem exists?

Kind regards
Christian

**Comment by JacobChrist** [comment]
Of course you have to rename the attached file to .sch

Ever file part in this BOM should have an attribute called PLP#, it only shows up in parts that have the attribute set in the part section of the file.


NAME	VALUE	PACKAGE	DESCRIPTION	DEVICE	OC_DIGIKEY	PLP	PLP#
C2	10uF	DISCRETE_C0805	Capacitor 0805				
C20	27pF	DISCRETE_C0603	Capacitor 0603			PLP-130	
C21	27pF	DISCRETE_C0603	Capacitor 0603			PLP-130	
C3	0.1uF	DISCRETE_C0603	Capacitor 0603			PLP-68	PLP-68
C4	0.1uF	DISCRETE_C0603	Capacitor 0603			PLP-68	PLP-68
C5	0.1uF	DISCRETE_C0603	Capacitor 0603			PLP-68	
C11	0.1uF	DISCRETE_C0603	Capacitor 0603			PLP-68	
C12	1uF	DISCRETE_C0603	Capacitor 0603			PLP-84	
C14	3.3uF	DISCRETE_C1210					
C34	68uF 50V	PLP-979	CAP ALUM 100UF 50V 20% SMD				
C35	100uF 16V	PLP-2	Capacitor 0603				
C13	470pF	DISCRETE_C0603	CAP CER 470PF 50V 5% C0G 0603				
C17	100uF 16V	PLP-2	Capacitor 0603				PLP-2
C15	100uF 16V	PLP-2	Capacitor 0603				PLP-2
C1	0.22uF	DISCRETE_C0603	Capacitor 0603				
C6	0.22uF	DISCRETE_C0603	Capacitor 0603				
C16	0.1uF 25V	DISCRETE_C0603	Capacitor 0603				
C18	10uF 6.3V	DISCRETE_C0603	Capacitor 0603				
C19	0.1uF 25V	DISCRETE_C0603	Capacitor 0603				
C22	10000pF 50V	DISCRETE_C0603	Capacitor 0603				
C23	47pF	DISCRETE_C0603	Capacitor 0603			PLP-132	
C24	47pF	DISCRETE_C0603	Capacitor 0603			PLP-132	
D1	600V 1A FAST	DIODE_SMA	DIODE				
D2	600V 1A FAST	DIODE_SMA	DIODE				
D32		SMA	3.0A SMD Schottky Barrier Rectifier				PLP-705
D31		SMA	3.0A SMD Schottky Barrier Rectifier				PLP-705
D9	70v 200mA	SOT-23	DIODE SCHOTTKY 70V 200MA SOT23				
IC6		TQFP64-10X10	PIC32MX440F256H edit this description				
J4		PLP-352	PLP-352				
J5		1751264	MKDS 1/ 4-3,5 Printklemme		277-1751-ND		PLP-714
J1		PLP-355	AMP connector				
J2		PLP-355	AMP connector				
J7		PLP-978	TERM BLOCK HDR 6POS VERT 3.5MM				
J6		PLP-355	AMP connector				
J3		PLP-737					
J8		PLP-737					
JP3		PLP-740	42 mil hole pattern with 16mil offset for jtag connector				
L5	1uH 960mA	PLP-835					
L9	150uH	SUMIDA_CDRH74					PLP-17
L1	220uH	MINE_0603				PLP-656	
LD1		DISCRETE_D0603	Discrete LED				
PROG		PLP-737					
Q1	60V 2.7A	SOT-23					
R19	22.0k	DISCRETE_R0603	RES 22.0K OHM 1/16W 1% 0603 SMD				PLP-32
R33	330	DISCRETE_R0603	Resistor			PLP-123	
R13	10K	DISCRETE_R0603	Resistor 10.0K 1% 0603				PLP-27
R14	100	DISCRETE_R0603	Resistor				
R16	100	DISCRETE_R0603	Resistor				
R17	100	DISCRETE_R0603	Resistor				
R37	200	DISCRETE_R0603	Resistor				
R42	100	DISCRETE_R0603	Resistor				
R3	10K	DISCRETE_R0603	Resistor 10.0K 1% 0603			PLP-27	PLP-27
R9	10K	DISCRETE_R0603	Resistor 10.0K 1% 0603			PLP-27	PLP-27
R35	3.09k	0603		0603			PLP-33
R36	1k	0603		0603			PLP-26
R46	0	DISCRETE_R0603	Resistor				PLP-91
R6	10K	DISCRETE_R0603	Resistor 10.0K 1% 0603				
R7	33K 5%	DISCRETE_R0603	Resistor				
R8	33K 5%	DISCRETE_R0603	Resistor				
R10	0	DISCRETE_R0603	Resistor				
R11	0	DISCRETE_R0603	Resistor				
R12	10k	DISCRETE_R0603	Resistor 10.0K 1% 0603				
R15	10K	DISCRETE_R0603	Resistor 10.0K 1% 0603				
R18	10K	DISCRETE_R0603	Resistor 10.0K 1% 0603				
R20	10K	DISCRETE_R0603	Resistor 10.0K 1% 0603				
R21	0	DISCRETE_R0603	Resistor				
R22	0	DISCRETE_R0603	Resistor				
R1	499K	DISCRETE_R0603	Resistor			PLP-273	
R23	10K	DISCRETE_R0603	Resistor 10.0K 1% 0603				
R2	330	DISCRETE_R0603	Resistor				
R4	10K	DISCRETE_R0603	Resistor 10.0K 1% 0603				
R24	330	DISCRETE_R0603	Resistor				
R25	330	DISCRETE_R0603	Resistor				
R26	10K	DISCRETE_R0603	Resistor 10.0K 1% 0603				
R5	10K	DISCRETE_R0603	Resistor 10.0K 1% 0603				
R27	10K	DISCRETE_R0603	Resistor 10.0K 1% 0603				
R28	10K	DISCRETE_R0603	Resistor 10.0K 1% 0603				
R29	120	DISCRETE_R0603	RES 120 OHM 1/10W 1% 0603 SMD				
U3		NATIONAL-SEMICONDUCTOR_SOIC-8	IC MULTI CONFIG ADJ 1.5A 8SOIC				
U4	3.3V 0.95A	SOT-223					
U2		PLP-969	IC MOTOR DRIVER SPI 36POWERSO				
U$3		PLP-288/712					
U$4	RS485	PLP-710	RS485 Half Duplex w/Tx Disable				
X1	FQ705B-8.000	DISCRETE_FQ7050B	Quartz Crystal			PLP-703	


[PLPB-0053.txt](https://github.com/I2SE/eagle-bom.py/files/65723/PLPB-0053.txt)

**Comment by JacobChrist** [comment]
If I add this code:
```python
        element = {}
        deviceset_tree = get_librarypart(drawing, elem.attrib['library'], elem.attrib['deviceset'])
        
        for technology_tree in deviceset_tree.iter('attribute'): #find('technologies/technology/attribute'):
            #print "tt", technology_tree.tag, technology_tree.attrib['name'] + "=" + technology_tree.attrib['value'],
            element[technology_tree.attrib['name']] = technology_tree.attrib['value']        

        element['NAME'] = elem.attrib['name']
``
into the bom_creation() then I get the result I need.  However, I don't think this is the proper solution.  All my parts have a single technology.  If they parts were built with more than one technology information would be lost, or incorrect information may show up in the BOM.  But his is getting closer.

Jacob

**Comment by JacobChrist** [comment]
Here is an example that uses technologies, but doesn't populate the PLP# column correctly with the fix given above (again, rename .txt to .sch):

[technologies.txt](https://github.com/I2SE/eagle-bom.py/files/67380/technologies.txt)

If you look at the schematic, you will see each part has a unique PLP# and value.

This is the resultant bom (with fix applied from above).  The PLP# column is populated, but incorrect except for the first part.  The PLP# column is completely missing when run from the head.

NAME,VALUE,PACKAGE,DESCRIPTION,DEVICE,PLP#
R1,0.0%,DISCRETE_R0603,,0.0OHM,PLP-91
R2,1.0K 0.1%,DISCRETE_R0603,,1.0K0.1%,PLP-91
R3,1.5K 1%,DISCRETE_R0603,,1.5K1%,PLP-91
R4,1K 1%,DISCRETE_R0603,,1K1%,PLP-91
R5,2.4K 1%,DISCRETE_R0603,,2.4K1%,PLP-91
R6,3.09K 1%,DISCRETE_R0603,,3.09K1%,PLP-91
R7,3K,DISCRETE_R0603,,3K5%,PLP-91
R8,7.5K 1%,DISCRETE_R0603,,7.5K1%,PLP-91
R9,10.0K 0.1%,DISCRETE_R0603,,10.0K0.1%,PLP-91
R10,10K 1%,DISCRETE_R0603,,10K1%,PLP-91
R11,12.4K 1%,DISCRETE_R0603,,12.4K1%,PLP-91
R12,22K 1%,DISCRETE_R0603,,22K1%,PLP-91
R13,100K 1%,DISCRETE_R0603,,100K1%,PLP-91
R14,100 0.1%,DISCRETE_R0603,,100OHM0.1%,PLP-91
R15,137K 1%,DISCRETE_R0603,,137K1%,PLP-91
R16,200 1%,DISCRETE_R0603,,200OHM1%,PLP-91

**Comment by JacobChrist** [comment]
I have a possible solution.  I will submit a pull request for review.

**Comment by JacobChrist** [comment]
I decided no to do a pull request, the test cases are not working and I'm too tired to fix tonight.  If you want to take a look at it here is my fork: https://github.com/pontech/eagle-bom.py/commits/master

**Comment by ch-aurich** [comment]
this should be fixed as well by merging #8
